Cleanup of RESP implementation

hderms opened the pull request:
* minimal integration tests with some additional debugging data if expectations don't match:
![new integration test error message](https://user-images.githubusercontent.com/833575/215856153-7384a3ba-2756-4bd8-a6ca-f42127a7bbe7.png)

* simple implementation of klog
* some `Display` trait implementations to make klogging easier

poppoerika commented:
@brayniac @eaddingtonwhite @danielamiao
Could you review this pr? Thank you!

eaddingtonwhite commented on src/rust/proxy/momento/Cargo.toml:
```diff
@@ -21,3 +21,17 @@ protocol-memcache = { path = "../../protocol/memcache" }
+	["../../../../config/momento_proxy.toml", "etc/momento_proxy/config.toml.example", "644"],	
+]
+maintainer-scripts = "debian/"
+systemd-units = { enable = false }
```
@brayniac I think maybe I messed this up in merging over changes from latest branch. Not sure why this is showing up in diff do you expect this change to be here.

brayniac replied:
This should not be here - I think I had reverted back to `c6a32f6` and rewrote the history.

eaddingtonwhite commented on src/rust/proxy/momento/debian/service:
```diff
@@ -0,0 +1,9 @@
+ExecStart=/usr/bin/momento_proxy /etc/momento_proxy/config.toml
+
+[Install]
+WantedBy=multi-user.target
```
@brayniac Same here Not sure why this is showing up in diff do you expect this change to be here.

brayniac replied:
This should not be here - you should be able to delete the `debian ` folder

poppoerika commented:
@brayniac - Removed the `debian` package.

brayniac committed:
stash rds

dependabot[bot] committed:
build(deps): bump tokio from 1.21.2 to 1.23.1 (#14)

Bumps [tokio](https://github.com/tokio-rs/tokio) from 1.21.2 to 1.23.1.
- [Release notes](https://github.com/tokio-rs/tokio/releases)
- [Commits](https://github.com/tokio-rs/tokio/compare/tokio-1.21.2...tokio-1.23.1)

---
updated-dependencies:
- dependency-name: tokio
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

brayniac committed:
feat: improve usability of memcache protocol (#15)

Improves usability of memcache protocol crate for use in rpc-perf

brayniac committed:
refactor resp protocol (#16)

Minor refactoring of the RESP protocol to make struct names more
consistent with memcache protocol implementation.

Adds constructor methods for request types to make it usable from
client code.

brayniac committed:
fix: address clippy warnings (#17)

Addresses clippy warnings

swlynch99 committed:
chore: update momento crate to v0.17.0 (#18)

Updates momento crate to 0.17.0

brayniac committed:
chore: update dependencies (#19)

Updates third-party dependencies to latest versions

brayniac committed:
docs: fix badges and links in readme (#20)

Fixes badges and links in readme to reflect 'main' as the default
branch.

brayniac committed:
docs: fix badges (#21)

One more fix for the badges in the readme

swlynch99 committed:
feat: support for the HINCRBY RESP command in Momento proxy (#22)

Adds support for HINCRBY RESP command in the RESP protocol and Momento proxy.

swlynch99 committed:
chore: Update momento to v0.21.0 (#30)

* chore: Update momento to v0.21.0
* chore: Install protoc in CI

hderms committed:
Cleanup:

fix clippy warnings

cleanup

cargo fmt

add Klog for RESP SET

add Klog for RESP GET

KLog support in RESP

enhance error output for more transparency

use helper methods/constants for responses

comment out memcached originated tests as they either depend on unimplemented RESP commands or are unclear if they'd work in Redis

remove test for non-redis command

fix typos in integration tests Segcache -> Rds

delete duplicate `set` implementation in comments (memcached copy/paste)

remove commented out buffer allocation line as GET in Redis can only fetch one key, unlike memcached

fix comment typo

hderms commented on src/protocol/resp/src/request/get.rs:
```diff
@@ -77,6 +78,19 @@ impl Compose for GetRequest {
+
+    fn klog(&self, response: &Self::Response) {
+        let code = match response {
+            Message::BulkString(_) if *response == Response::null() => MISS,
```
this is an area where having storage return an enum of possible result conditions would be nice

hderms commented on src/protocol/resp/src/request/mod.rs:
```diff
@@ -36,6 +38,11 @@ pub use hmget::*;
 pub use hset::*;
 pub use hvals::*;
 pub use set::*;
+// response codes for klog
```
I used the same codes as memcached but I'll defer to you if you'd prefer something more redis-y

hderms commented on src/protocol/resp/src/request/mod.rs:
```diff
@@ -321,3 +333,23 @@ pub enum ExpireTime {
+impl Display for ExpireTime {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ExpireTime::Seconds(s) => write!(f, "{}s", s),
+            ExpireTime::Milliseconds(ms) => write!(f, "{}ms", ms),
+            ExpireTime::UnixSeconds(s) => write!(f, "{}unix_secs", s),
+            ExpireTime::UnixMilliseconds(ms) => write!(f, "{}unix_ms", ms),
+            ExpireTime::KeepTtl => write!(f, "keep_ttl"),
```
could see the formatting of these being an issue

hderms committed:
get it compiling

brayniac committed:
Cleanup entrystore for resp

removes commented out code and fixes copyright

brayniac committed:
Cleanup RESP storage trait

Remove commented functions

brayniac committed:
Fix copyright header

brayniac committed:
Fix copyright header

brayniac committed:
Cleanup integration test

hderms committed:
remove redis dump file, errantly committed

swlynch99 committed:
Implement resp SADD command (#32)

hderms committed:
update Rds benchmarks to compile/complete successfully

hderms committed:
Fix comments that errantly refer to segcache

hderms committed:
update licenses to Pelikan

hderms committed:
correct brian's email

hderms committed:
chore: update Brian's email in all toml files (#34)

Update Brian's email address.

hderms committed:
Remove errant license changes and remove commented out code

hderms committed:
remove more errant copyright changes

hderms committed:
cargo fmt

swlynch99 committed:
chore: update momento crate requirement to the latest version (#35)

* chore: update momento crate requirement to the latest version
* chore: bump momento to 0.21.9

swlynch99 committed:
chore: Implement the LINDEX command (#36)

swlynch99 committed:
feat: Implement resp SREM command (#37)

swlynch99 committed:
feat: Implement resp LLEN command (#38)

swlynch99 committed:
chore: fix clippy warnings (#48)

swlynch99 committed:
chore: refactor momento frontend to reduce code duplication (#42)

The command handling in the momento proxy has a _lot_ of duplicated code (about 60% of lines for each new command are copy-paste). By refactoring things I have moved the common code up one level so that common error handling and metrics work no longer needs to be copy-pasted in to each new function.

The detailed changes are as follows:
- Introduce a new error type that covers all error cases encountered in the resp client.
- Move common metric updates up into `handle_resp_client` using the error type above to figure out what needs to be updated.
- Update `GET` and `HDEL` to take advantage of this as an example of the benefits.

This commit does not update all the commands. That will happen in a future set of commits.

swlynch99 committed:
feat: implement resp SDIFF command (#43)

hderms committed:
fix: issue #40 panic RESP protocol parser

Fixes issue #40 by correcting the logic when a LF is received instead of a CRLF.
Adds unit tests.

---------

Co-authored-by: Sean Lynch <42618346+swlynch99@users.noreply.github.com>

swlynch99 committed:
feat: Implement resp SUNION command (#44)

swlynch99 committed:
feat: Implement resp SINTER command (#45)

swlynch99 committed:
feat: Implement resp `SMEMBERS` command (#46)

swlynch99 committed:
feat: Implement resp `SISMEMBER` command (#47)

swlynch99 committed:
feat: Implement resp `LPOP` command (#49)

swlynch99 committed:
feat: Implement resp LRANGE command (#52)

brayniac committed:
Merge branch 'main' into redis

brayniac committed:
Update rds.toml

Change log file names in example config

hderms committed:
change comments

swlynch99 committed:
feat: Implement resp `LPUSH` command (#53)

hderms committed:
match memcache klog

hderms committed:
fix comments/names

hderms committed:
clone SegcacheConfig into RdsConfig and use it

swlynch99 committed:
feat: Implement resp RPUSH command (#54)

hderms committed:
use NotStored when we don't return a simple string for SET

swlynch99 committed:
feat: Implement resp LTRIM command (#55)

swlynch99 committed:
chore: Add check-success jobs to end of CI workflows (#60)

Check-success jobs for CI to be used in branch protection rules.

swlynch99 committed:
fix: ignore flaky integration test on MacOS (#58)

Co-authored-by: Brian Martin <brian@pelikan.io>

swlynch99 committed:
feat: Implement resp RPOP command (#50)

hderms committed:
Merge branch 'main' into redis
